Optimize SVGTextLayoutAttributesBuilder rotation attribute handling

Ahmad-S792 opened the pull request:
#### 27a74ff8ea3eb998f9c751d749e1097096e80381
<pre>
Optimize SVGTextLayoutAttributesBuilder rotation attribute handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=303279">https://bugs.webkit.org/show_bug.cgi?id=303279</a>
<a href="https://rdar.apple.com/165585366">rdar://165585366</a>

Reviewed by Brent Fulgham.

The fillCharacterDataMap() method was using a two-pass approach for
handling rotation attributes: first iterating through explicit rotate
values, then a second loop to propagate the last rotation value to
remaining characters.
This patch optimizes the rotation handling to use a single-pass
approach by leveraging std::min() to automatically select the last
rotation value when the index exceeds the list size.

Also removes the now-unused updateCharacterData() helper function that
was specific to the old two-pass implementation.

Inspired by: <a href="https://chromium.googlesource.com/chromium/src.git/+/d934544befc4375078bc94e24234a98c60c55b1a">https://chromium.googlesource.com/chromium/src.git/+/d934544befc4375078bc94e24234a98c60c55b1a</a>

* Source/WebCore/rendering/svg/SVGTextLayoutAttributesBuilder.cpp:
(WebCore::SVGTextLayoutAttributesBuilder::fillCharacterDataMap):

Canonical link: <a href="https://commits.webkit.org/304016@main">https://commits.webkit.org/304016@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dcb65a5a9ae477472e00ee016369ba109715409a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134299 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/159/builds/6801 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45520 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/141871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/86332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/87dacedf-f827-494f-bf69-6bfb30e58a46) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136169 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7383 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6664 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/141871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/86332 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/4b93158b-8a1a-4ae3-b2db-e8cce945b4a2) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137246 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/5145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120376 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/141871 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/2629ff2c-981a-457d-a324-803e2a5ed0b5) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/5019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2272 "Failed to checkout and rebase branch from PR 54597") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/114218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38514 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144525 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6474 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39094 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111071 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/160/builds/6556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/5445 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111325 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28237 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-26-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116645 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60260 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6526 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/157/builds/6359 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70083 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/6596 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/152/builds/6472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/rendering/svg/SVGTextLayoutAttributesBuilder.cpp:
```diff
@@ -195,32 +195,33 @@ void SVGTextLayoutAttributesBuilder::fillCharacterDataMap(const TextPosition& po
+        if (rotateListPtr) {
+            unsigned rotateIndex = std::min(i, rotateListSize - 1);
+            data.rotate = rotateList.items()[rotateIndex]->value();
+        }
```
Aren't lines 212 through 224 just the implementation of `updateCharacterData`? Maybe we could just call that helper function instead of repeating it here?

(Or maybe my eyesight is bad and I missed a detail?)

Ahmad-S792 replied:
@brentfulgham - Correct and it was kind of similar. However, updateCharacterData() is actually only used in the old implementation and becomes dead code with this optimization. I've removed it as part of this patch now since the new single-pass approach needs different index handling for rotation vs other attributes.

brentfulgham reviewed (approved):
LGTM. I think we can just call the helper function, though.

brentfulgham reviewed (approved):
Looks like a good cleanup. r=me

Ahmad-S792 committed:
Optimize SVGTextLayoutAttributesBuilder rotation attribute handling
https://bugs.webkit.org/show_bug.cgi?id=303279
rdar://165585366

Reviewed by Brent Fulgham.

The fillCharacterDataMap() method was using a two-pass approach for
handling rotation attributes: first iterating through explicit rotate
values, then a second loop to propagate the last rotation value to
remaining characters.
This patch optimizes the rotation handling to use a single-pass
approach by leveraging std::min() to automatically select the last
rotation value when the index exceeds the list size.

Also removes the now-unused updateCharacterData() helper function that
was specific to the old two-pass implementation.

Inspired by: https://chromium.googlesource.com/chromium/src.git/+/d934544befc4375078bc94e24234a98c60c55b1a

* Source/WebCore/rendering/svg/SVGTextLayoutAttributesBuilder.cpp:
(WebCore::SVGTextLayoutAttributesBuilder::fillCharacterDataMap):

Canonical link: https://commits.webkit.org/304016@main
